File: quote/Projects/PluginModelIO/CommonLoader.vb
Imports System.Data.SqlClient
Imports System.Reflection
Imports System.Data.OleDb
Imports System.Transactions

Imports Model.Common

Imports DB.QuoteDataBase
Imports DB.QuoteDataBaseTableAdapters

Public Class CommonLoader

    Public Shared Sub LoadNoteProperties(ByVal id As Integer, _
                                         ByVal obj As Object)
        LoadProperties(id, CommonSaver.NOTE_PROPERTIES_ID, obj)
    End Sub

    Public Shared Sub LoadOtherProperties(ByVal id As Integer, _
                                          ByVal obj As Object)
        LoadProperties(id, CommonSaver.OTHER_PROPERTIES_ID, obj)
    End Sub

    Public Shared Sub LoadComputationProperties(ByVal id As Integer, _
                                                ByVal obj As Object)
        LoadProperties(id, CommonSaver.COMPUTATION_PROPERTIES_ID, obj)
    End Sub

    Public Shared Sub LoadComponents(ByVal q As Model.Common.Header)

        Dim adaptor As New _QuoteDetailTableAdapter
        Dim partAdaptor As New WireComponentSourceTableAdapter
        Dim wireAdaptor As New WireSourceTableAdapter
        Dim gageAdaptor As New GageTableAdapter
        Dim id As Integer = q.PrimaryProperties.CommonId
        Dim table As _QuoteDetailDataTable = adaptor.GetDataByQuoteID(id)
        For Each row As _QuoteDetailRow In table.Rows

            Dim detail As Model.Common.Detail = Nothing

            Dim parts As WireComponentSourceDataTable
            parts = partAdaptor.GetDataByPartNumber(row.ProductCode)
            If (parts.Count > 0) Then
                Dim part As WireComponentSourceRow
                part = parts(0)
                Dim price As Decimal = 0
                If Not part.IsQuotePriceNull Then
                    price = part.QuotePrice
                End If
                
                Dim data As ProductBuildData = ProductDB.Load( _
                    part.PartNumber, price, _
                    0, False, Nothing, part, "", 0)
                Dim partObj As New Model.Product(Data)

                detail = q.NewDetail(partObj)
            End If

            Dim wires As WireSourceDataTable
            wires = wireAdaptor.GetDataByPartNumber(row.ProductCode)
            If (wires.Count > 0) Then
                Dim wire As WireSourceRow
                wire = wires(0)
                Dim gage As String = ""
                Dim gageTable As GageDataTable
                gageTable = gageAdaptor.GetDataByGageID(wire.GageID)
                If gageTable IsNot Nothing Then
                    Dim gageRow As GageRow = gageTable.Rows(0)
                    gage = gageRow.Gage
                End If

                Dim price As Decimal = 0
                If Not wire.IsQuotePriceNull Then
                    price = wire.QuotePrice
                End If
                
                Dim data As ProductBuildData = ProductDB.Load( _
                    wire.PartNumber, price, _
                    gage, True, wire, Nothing, "", 0)
                Dim wireObj As New Model.Product(Data)

                detail = q.NewDetail(wireObj)
            End If

            If (detail IsNot Nothing) Then
                detail.Qty = row.Qty
                LoadProperties(id, row.id, detail.QuoteDetailProperties)
            End If
        Next
    End Sub

    Public Shared Sub LoadProperties(ByVal id As Integer, _
                                     ByVal childId As Integer, _
                                     ByVal obj As Object)

        Dim props As PropertyInfo() = obj.GetType.GetProperties
        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter
        For Each p As PropertyInfo In props

            Dim table As _QuotePropertiesDataTable = _
                adaptor.GetDataByQuoteAndName(id, childId, p.Name)

            If table.Rows.Count > 0 Then
                Dim row As _QuotePropertiesRow = table.Rows(0)
                If Not row.IsPropertyStringValueNull Then
                    If p.PropertyType.Name = "String" And p.CanWrite Then
                        p.SetValue(obj, row.PropertyStringValue, Nothing)
                    End If
                End If
                If Not row.IsPropertyStringValueNull Then
                    If p.PropertyType.Name = "Boolean" And p.CanWrite Then
                        If row.PropertyStringValue = "True" Then
                            p.SetValue(obj, True, Nothing)
                        Else
                            p.SetValue(obj, False, Nothing)
                        End If
                    End If
                End If
                If Not row.IsPropertyIntegerValueNull Then
                    If p.PropertyType.Name = "Int32" And p.CanWrite Then
                        p.SetValue(obj, row.PropertyIntegerValue, Nothing)
                    End If
                End If
                If Not row.IsPropertyDecimalValueNull Then
                    If p.PropertyType.Name = "Decimal" And p.CanWrite Then
                        p.SetValue(obj, row.PropertyDecimalValue, Nothing)
                    End If
                    If p.PropertyType.Name = "Decimal" And Not p.CanWrite Then
                        Dim m As MethodInfo = obj.GetType.GetMethod("Set" + p.Name)
                        If (m IsNot Nothing) Then
                            m.Invoke(obj, New Object() {row.PropertyDecimalValue})
                        End If
                    End If
                End If
                If Not row.IsPropertyDateValueNull Then
                    If p.PropertyType.Name = "DateTime" And p.CanWrite Then
                        Dim dt As DateTime = row.PropertyDateValue
                        If dt.Year > 1900 Then
                            p.SetValue(obj, row.PropertyDateValue, Nothing)
                        End If
                    End If
                End If
            End If
        Next
    End Sub

End Class

File: trunk/quote/Projects/PluginModelIO/CommonLoader.cs
namespace Model.IO
{
    using System;
    using System.Linq;
    using System.Reflection;
    
    using DB;
    using DB.QuoteDataBaseTableAdapters;

    public static class CommonLoader
    {
        public static void LoadNoteProperties(int id, object target)
        {
            LoadProperties(id, CommonSaver.NotePropertiesId, target);
        }

        public static void LoadOtherProperties(int id, object target)
        {
            LoadProperties(id, CommonSaver.OtherPropertiesId, target);
        }

        public static void LoadComputationProperties(int id, object target)
        {
            LoadProperties(id, CommonSaver.ComputationPropertiesId, target);
        }

        public static void LoadComponents(Model.Common.Header header)
        {
            _QuoteDetailTableAdapter adaptor = new _QuoteDetailTableAdapter();
            WireComponentSourceTableAdapter partAdaptor = new WireComponentSourceTableAdapter();
            WireSourceTableAdapter wireAdaptor = new WireSourceTableAdapter();
            GageTableAdapter gageAdaptor = new GageTableAdapter();
            
            int id = header.PrimaryProperties.CommonId;
            
            DB.QuoteDataBase._QuoteDetailDataTable table = adaptor.GetDataByQuoteID(id);

            foreach (DB.QuoteDataBase._QuoteDetailRow row in table.Rows) 
            {
                Model.Common.Detail detail = null;

                DB.QuoteDataBase.WireComponentSourceDataTable parts = null;
                parts = partAdaptor.GetDataByPartNumber(row.ProductCode);
                
                if (parts.Count > 0)
                {
                    DB.QuoteDataBase.WireComponentSourceRow part = null;
                    part = parts[0];
                    decimal price = 0;
                    
                    if (!part.IsQuotePriceNull()) 
                    {
                        price = part.QuotePrice;
                    }

                    Model.Product partObj = ProductDB.Load(
                        part.PartNumber, 
                        price, 
                        string.Empty, 
                        false, 
                        null, 
                        part, 
                        string.Empty, 
                        0);

                    detail = header.NewDetail(partObj);
                }

                DB.QuoteDataBase.WireSourceDataTable wires = null;
                wires = wireAdaptor.GetDataByPartNumber(row.ProductCode);
                
                if (wires.Count > 0)
                {
                    DB.QuoteDataBase.WireSourceRow wire = null;
                    wire = wires[0];
                    
                    string gage = string.Empty;
                    DB.QuoteDataBase.GageDataTable gageTable = null;
                    
                    gageTable = gageAdaptor.GetDataByGageID(wire.GageID);
                    if (gageTable != null) 
                    {
                        DB.QuoteDataBase.GageRow gageRow = gageTable.Rows[0] as DB.QuoteDataBase.GageRow;
                        gage = gageRow.Gage;
                    }

                    decimal price = 0;
                    if (!wire.IsQuotePriceNull()) 
                    {
                        price = wire.QuotePrice;
                    }

                    Model.Product wireObj = ProductDB.Load(
                        wire.PartNumber, 
                        price, 
                        gage, 
                        true, 
                        wire, 
                        null, 
                        string.Empty, 
                        0);

                    detail = header.NewDetail(wireObj);
                }

                if (detail != null)
                {
                    detail.Qty = row.Qty;
                    LoadProperties(id, (int)row.id, detail.QuoteDetailProperties);
                }
            }
        }

        public static void LoadProperties(
            int id, 
            int childId, 
            object target)
        {
            PropertyInfo[] props = target.GetType().GetProperties();
            
            DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter();

            foreach (PropertyInfo p in props) 
            {
                DB.QuoteDataBase._QuotePropertiesDataTable table = 
                    adaptor.GetDataByQuoteAndName(id, childId, p.Name);

                if (table.Rows.Count > 0) 
                {
                    DB.QuoteDataBase._QuotePropertiesRow row = 
                        table.Rows[0] as DB.QuoteDataBase._QuotePropertiesRow;
                    
                    if (!row.IsPropertyStringValueNull()) 
                    {
                        if (p.PropertyType.Name == "String" & p.CanWrite) 
                        {
                            p.SetValue(target, row.PropertyStringValue, null);
                        }
                    }
                    
                    if (!row.IsPropertyStringValueNull()) 
                    {
                        if (p.PropertyType.Name == "Boolean" & p.CanWrite) 
                        {
                            if (row.PropertyStringValue == "True") 
                            {
                                p.SetValue(target, true, null);
                            } 
                            else 
                            {
                                p.SetValue(target, false, null);
                            }
                        }
                    }
                    
                    if (!row.IsPropertyIntegerValueNull()) 
                    {
                        if (p.PropertyType.Name == "Int32" & p.CanWrite) 
                        {
                            p.SetValue(target, row.PropertyIntegerValue, null);
                        }
                    }
                    
                    if (!row.IsPropertyDecimalValueNull()) 
                    {
                        if (p.PropertyType.Name == "Decimal" & p.CanWrite) 
                        {
                            p.SetValue(target, row.PropertyDecimalValue, null);
                        }
                        
                        if (p.PropertyType.Name == "Decimal" & !p.CanWrite) 
                        {
                            MethodInfo m = target.GetType().GetMethod("Set" + p.Name);
                            if (m != null)
                            {
                                m.Invoke(target, new object[] { row.PropertyDecimalValue });
                            }
                        }
                    }
                    
                    if (!row.IsPropertyDateValueNull()) 
                    {
                        if (p.PropertyType.Name == "DateTime" & p.CanWrite) 
                        {
                            DateTime dt = row.PropertyDateValue;
                            if (dt.Year > 1900) 
                            {
                                p.SetValue(target, row.PropertyDateValue, null);
                            }
                        }
                    }
                }
            }
        }
    }
}
